docs/data/data-grid/server-side-data/index.md

```diff
@@ -0,0 +1,269 @@
+    truncated?: number; // optional: to reflect rowCount is inaccurate (will trigger `x-y of many` in pagination)
+  };
+}
+```
+
+#### Updating data
+
+If provided, the method `dataSource.updateRow` will be called with the `GridRowModel` object whenever the user edits a row. This method is optional and you can skip it if you don't need to update the data on the server. It will work in a similar way as the `processRowUpdate` prop.
+
+#### Data grid props
+
+These data grid props will work with the server-side data source:
+
+- `dataSource: DataSource`: the data source object that you need to implement
+- `queryClient: QueryClient`: optional: if not provided, a new instance of `QueryClient` will be created internally
+- `rows`: will be ignored, could be skipped when `dataSource` is provided
+- `rowCount`: will be used to identify the total number of rows in the grid, if not provided, the grid will use the _GetRowsResponse.rowCount_ value
+
+Props related to grouped data (`treeData` and `rowGrouping`):
+
+- `getGroupKey(row: GridRowModel): string`
+
+  will be used by the grid to group rows by their parent group
+  This effectively replaces `getTreeDataPath`.
+  Consider this structure:
+
+  ```
+  - (1) Sarah // groupKey 'Sarah'
+    - (2) Thomas // groupKey 'Thomas'
+  ```
+
+  When (2) is expanded, the `getRows` function will be called with group keys `['Sarah', 'Thomas']`.
+
+- `hasChildren?(row: GridRowModel): boolean`
+
+  Will be used by the grid to determine if a row has children on server
+
+- `getChildrenCount?: (row: GridRowModel) => number`
+
+  Will be used by the grid to determine the number of children of a row on server
+
+#### Existing server-side features
+
+The server-side `dataSource` will change a bit the way existing server-side features work. Whenever a valid data source is passed the features `filtering`, `sorting`, `pagination` will automatically be set to `server`.
+
+However, if for some reason, you want to explicitly work one or more of these features to work on `client`, you can do so by passing the `client` value to the feature. Here's an example:
+
+```tsx
+<DataGridPro
+  columns={columns}
+  dataSource={myCustomDataSource} // will set `filtering`, `sorting`, `pagination` to `server`
+  filterMode="client" // will set `filtering` to `client` i.e. data won't be refetched on filter change
+/>
+```
+
+#### Caching
+
+The data grid will cache the data it receives from the server. This means that if the user navigates to a page that has already been fetched, the grid will not call the `getRows` function again. This is to avoid unnecessary calls to the server.
+
+By default, the data grid will use the tanstack query library to cache the data. A `QueryClient` will be created internally and used with the data source. You can also pass your own instance of `QueryClient` with a optional custom configuration and use it to play with the cached data. You can also use this client to invalidate the cache when you need to.
```

Comment:
```suggestion
By default, the data grid will use the `@tanstack/react-query` library to cache the data. A `QueryClient` will be created internally and used with the data source. You can also pass your own instance of `QueryClient` with a optional custom configuration and use it to play with the cached data. You can also use this client to invalidate the cache when you need to.
```

Reply:
For my information, why did we decide to use react-query?

Reply:
To take advantage of the powerful caching functionality it provides and allow users the customization with caching that it provides.

Reply:
We could also think of making this library agnostic in future so that users are free to use:
- No cache
- custom made cache
- react-query
- swr
- any other that they want

But I feel that should happen if there's a strong need for this kind of customization and people are demanding it. As far as I think, most of the caching use cases could be covered with tanstack query.

Reply:
> we expose an external library as an API so we're a bit stuck with it

That's a good point. Also, we'll only use a part of the library and I'm not sure how well will tree shaking work in this case.

> For something that in the end doesn't seem terribly more than just a Map<requestId, Response>.

Yes, plus cache configuration and proper invalidation.

I considered using react query as a shortcut, but now I feel more convinced to our own implementation 🙂

Reply:
I'm still not convinced.

I think we need that for CSS because users come in with their styling solution already. It doesn't apply for fetching clients. And if we expose a library-agnostic interface, then it makes absolutely no difference for users what the underlying client is.

But I'll note that if we could expose an opaque or library-agnostic interface, then I wouldn't really mind using react-query internally for the next release - because we would still have the liberty of improving it anytime.

What are the use-cases we're trying to solve by exposing that object? Do you have any pseudo-code or doc examples I could look at to understand better?

Reply:
> What are the use-cases we're trying to solve by exposing that object? Do you have any pseudo-code or doc examples I could look at to understand better?

I'm sure there could be multiple ones, but one of the prominent use-cases we were targeting is the easy synchronization with the other apps like toolpad or charts.
An easy way out that cache from the library can provide is being able to subscribe to the changes and get the updated values from the cache like shown in this example: https://codesandbox.io/s/datagridpro-basic-forked-49zpsy?file=/src/App.tsx (but yes, that could be done with the custom implementation)

> But I'll note that if we could expose an opaque or library-agnostic interface, then I wouldn't really mind using react-query internally for the next release - because we would still have the liberty of improving it anytime.

Thanks for sharing the useful thoughts on this. I didn't think like this previously, I seem to agree with you that we shouldn't expose the interface of an external library we don't control. And I believe it might get a bit trickier to nail, but we can solve the problems like cache data sharing and request deduplication internally.
I'll update the documentation and see how it looks like for an end user.

As for the customization the library provided, in the beginning, we can even chose to not allowing any customization on the user's side, to follow it up later as per the needs of the users.

Reply:
I've removed the implementation details and interface around cache for now, currently decision is to not directly expose a library and try to do internal implementation for caching. Resolving the conversation, feel free to unresolved if some point still needs discussion.